File: flutter_app/native_modules/android/HeartRateMonitor.kt
// HeartRateMonitor.kt
// Titan Native Module - Android BLE Heart Rate Monitor
// Handles: BLE connection to HR monitors, background streaming, battery optimization

package com.titan.app.native_modules

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.ParcelUuid
import androidx.core.content.ContextCompat
import io.flutter.plugin.common.MethodChannel
import java.util.UUID

object HeartRateUUIDs {
    val HEART_RATE_SERVICE = UUID.fromString("0000180d-0000-1000-8000-00805f9b34fb")
    val HEART_RATE_MEASUREMENT = UUID.fromString("00002a37-0000-1000-8000-00805f9b34fb")
    val BATTERY_SERVICE = UUID.fromString("0000180f-0000-1000-8000-00805f9b34fb")
    val BATTERY_LEVEL = UUID.fromString("00002a19-0000-1000-8000-00805f9b34fb")
    val CCC_DESCRIPTOR = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
}

data class HeartRateReading(
    val bpm: Int,
    val hrv: Double?,
    val source: String,
    val timestamp: Long,
    val batteryLevel: Int?
)

data class HeartRateDevice(
    val device: BluetoothDevice,
    var name: String,
    var rssi: Int,
    var batteryLevel: Int? = null,
    var isConnected: Boolean = false
)

class HeartRateMonitor(private val context: Context) {
    
    companion object {
        const val BATCH_SIZE = 10
    }
    
    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager.adapter
    private val bluetoothLeScanner = bluetoothAdapter?.bluetoothLeScanner
    
    private var isScanning = false
    private var isMonitoring = false
    private var discoveredDevices: MutableMap<String, HeartRateDevice> = mutableMapOf()
    private var connectedDevice: HeartRateDevice? = null
    private var bluetoothGatt: BluetoothGatt? = null
    
    private var readingBuffer: MutableList<HeartRateReading> = mutableListOf()
    private var methodChannel: MethodChannel? = null
    
    // MARK: - Flutter Channel Setup
    
    fun setMethodChannel(channel: MethodChannel) {
        methodChannel = channel
    }
    
    // MARK: - Scanning
    
    fun startScan() {
        if (isScanning || bluetoothLeScanner == null) return
        
        if (!hasBluetoothPermission()) {
            methodChannel?.invokeMethod("onError", "Bluetooth permission not granted")
            return
        }
        
        discoveredDevices.clear()
        isScanning = true
        
        val filters = listOf(
            ScanFilter.Builder()
                .setServiceUuid(ParcelUuid(HeartRateUUIDs.HEART_RATE_SERVICE))
                .build()
        )
        
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()
        
        bluetoothLeScanner.startScan(filters, settings, scanCallback)
    }
    
    fun stopScan() {
        if (!isScanning || bluetoothLeScanner == null) return
        bluetoothLeScanner.stopScan(scanCallback)
        isScanning = false
    }
    
    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val deviceId = device.address
            
            val hrDevice = HeartRateDevice(
                device = device,
                name = device.name ?: "Unknown Device",
                rssi = result.rssi
            )
            
            discoveredDevices[deviceId] = hrDevice
            
            methodChannel?.invokeMethod("onDeviceDiscovered", mapOf(
                "id" to deviceId,
                "name" to hrDevice.name,
                "rssi" to hrDevice.rssi
            ))
        }
        
        override fun onScanFailed(errorCode: Int) {
            methodChannel?.invokeMethod("onError", "Scan failed with error: $errorCode")
        }
    }
    
    // MARK: - Connection
    
    fun connect(deviceId: String) {
        val device = discoveredDevices[deviceId] ?: return
        
        if (!hasBluetoothPermission()) return
        
        stopScan()
        
        bluetoothGatt = device.device.connectGatt(
            context,
            false,
            gattCallback,
            BluetoothDevice.TRANSPORT_LE
        )
    }
    
    fun disconnect() {
        bluetoothGatt?.close()
        bluetoothGatt = null
        connectedDevice = null
    }
    
    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    gatt.discoverServices()
                    methodChannel?.invokeMethod("onDeviceConnected", mapOf(
                        "id" to gatt.device.address,
                        "name" to connectedDevice?.name ?: "Unknown"
                    ))
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    methodChannel?.invokeMethod("onDeviceDisconnected", mapOf(
                        "id" to gatt.device.address
                    ))
                }
            }
        }
        
        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) return
            
            // Subscribe to heart rate measurement
            val hrService = gatt.getService(HeartRateUUIDs.HEART_RATE_SERVICE)
            hrService?.getCharacteristic(HeartRateUUIDs.HEART_RATE_MEASUREMENT)?.let { char ->
                enableNotification(gatt, char)
            }
            
            // Read battery level
            val batteryService = gatt.getService(HeartRateUUIDs.BATTERY_SERVICE)
            batteryService?.getCharacteristic(HeartRateUUIDs.BATTERY_LEVEL)?.let { char ->
                gatt.readCharacteristic(char)
            }
        }
        
        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            if (characteristic.uuid == HeartRateUUIDs.HEART_RATE_MEASUREMENT) {
                val bpm = parseHeartRate(characteristic)
                if (bpm > 0 && isMonitoring) {
                    processHeartRate(bpm)
                }
            }
        }
        
        override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            if (characteristic.uuid == HeartRateUUIDs.BATTERY_LEVEL) {
                val battery = characteristic.value?.firstOrNull()?.toInt() ?: 0
                connectedDevice?.batteryLevel = battery
                methodChannel?.invokeMethod("onBatteryLevelUpdate", battery)
            }
        }
    }
    
    private fun enableNotification(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
        gatt.setCharacteristicNotification(characteristic, true)
        
        val descriptor = characteristic.getDescriptor(HeartRateUUIDs.CCC_DESCRIPTOR)
        descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
        gatt.writeDescriptor(descriptor)
    }
    
    // MARK: - Data Processing
    
    private fun parseHeartRate(characteristic: BluetoothGattCharacteristic): Int {
        val flags = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0)
        val is16Bit = (flags and 0x01) != 0
        
        return if (is16Bit) {
            characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, 1)
        } else {
            characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 1)
        }
    }
    
    private fun processHeartRate(bpm: Int) {
        val reading = HeartRateReading(
            bpm = bpm,
            hrv = null,
            source = connectedDevice?.name ?: "BLE",
            timestamp = System.currentTimeMillis(),
            batteryLevel = connectedDevice?.batteryLevel
        )
        
        readingBuffer.add(reading)
        
        methodChannel?.invokeMethod("onHeartRateUpdate", mapOf(
            "bpm" to bpm,
            "timestamp" to reading.timestamp
        ))
        
        if (readingBuffer.size >= BATCH_SIZE) {
            emitBuffer()
        }
    }
    
    private fun emitBuffer() {
        if (readingBuffer.isEmpty()) return
        
        val batch = readingBuffer.map { reading ->
            mapOf(
                "bpm" to reading.bpm,
                "hrv" to reading.hrv,
                "source" to reading.source,
                "timestamp" to reading.timestamp,
                "batteryLevel" to reading.batteryLevel
            )
        }
        
        methodChannel?.invokeMethod("onHeartRateBatch", mapOf("readings" to batch))
        readingBuffer.clear()
    }
    
    // MARK: - Monitoring Control
    
    fun startMonitoring() {
        isMonitoring = true
        readingBuffer.clear()
    }
    
    fun stopMonitoring() {
        isMonitoring = false
        emitBuffer()
    }
    
    // MARK: - Helpers
    
    private fun hasBluetoothPermission(): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }
    }
    
    fun getDiscoveredDevices(): List<Map<String, Any>> {
        return discoveredDevices.values.map { device ->
            mapOf(
                "id" to device.device.address,
                "name" to device.name,
                "rssi" to device.rssi,
                "isConnected" to device.isConnected
            )
        }
    }
    
    fun getConnectedDeviceInfo(): Map<String, Any>? {
        return connectedDevice?.let { device ->
            mapOf(
                "id" to device.device.address,
                "name" to device.name,
                "batteryLevel" to device.batteryLevel
            )
        }
    }
    
    fun getLatestHeartRate(): Map<String, Any>? {
        return readingBuffer.lastOrNull()?.let { reading ->
            mapOf(
                "bpm" to reading.bpm,
                "timestamp" to reading.timestamp
            )
        }
    }
    
    fun isScanning(): Boolean = isScanning
    fun isMonitoring(): Boolean = isMonitoring
}

// Flutter Plugin
class HeartRateMonitorPlugin(private val context: Context) : MethodChannel.MethodCallHandler {
    
    private val monitor = HeartRateMonitor(context)
    private var methodChannel: MethodChannel? = null
    
    fun setMethodChannel(channel: MethodChannel) {
        methodChannel = channel
        monitor.setMethodChannel(channel)
        channel.setMethodCallHandler(this)
    }
    
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startScan" -> { monitor.startScan(); result.success(null) }
            "stopScan" -> { monitor.stopScan(); result.success(null) }
            "getDiscoveredDevices" -> result.success(monitor.getDiscoveredDevices())
            "connect" -> { monitor.connect(call.argument<String>("deviceId")!!); result.success(null) }
            "disconnect" -> { monitor.disconnect(); result.success(null) }
            "getConnectedDevice" -> result.success(monitor.getConnectedDeviceInfo())
            "startMonitoring" -> { monitor.startMonitoring(); result.success(null) }
            "stopMonitoring" -> { monitor.stopMonitoring(); result.success(null) }
            "getLatestHeartRate" -> result.success(monitor.getLatestHeartRate())
            "isMonitoring" -> result.success(monitor.isMonitoring())
            else -> result.notImplemented()
        }
    }
}

File: flutter_app/native_modules/ios/HeartRateMonitor.swift
//
//  HeartRateMonitor.swift
//  Titan Native Module - BLE Heart Rate Monitor
//
//  Handles: BLE connection to heart rate monitors, background streaming, battery optimization
//  Uses: CoreBluetooth, Heart Rate Profile (UUID 0x180D)
//

import CoreBluetooth
import Flutter
import UIKit

// MARK: - Constants

struct HeartRateServiceUUID {
    static let heartRate = CBUUID(string: "180D")
    static let heartRateMeasurement = CBUUID(string: "2A37")
    static let bodySensorLocation = CBUUID(string: "2A38")
    static let heartRateControlPoint = CBUUID(string: "2A39")
    
    static let batteryService = CBUUID(string: "180F")
    static let batteryLevel = CBUUID(string: "2A19")
    
    static let deviceInfoService = CBUUID(string: "180A")
    static let manufacturerName = CBUUID(string: "2A29")
    static let modelNumber = CBUUID(string: "2A24")
}

// MARK: - Data Models

struct HeartRateReading: Codable {
    let bpm: Int
    let hrv: Double?  // Heart Rate Variability (if available)
    let source: String  // Device name
    let timestamp: Date
    let batteryLevel: Int?
    
    func toDictionary() -> [String: Any] {
        return [
            "bpm": bpm,
            "hrv": hrv ?? NSNull(),
            "source": source,
            "timestamp": timestamp.timeIntervalSince1970 * 1000,
            "batteryLevel": batteryLevel ?? NSNull()
        ]
    }
}

struct HeartRateDevice {
    let peripheral: CBPeripheral
    var name: String
    var rssi: Int
    var batteryLevel: Int?
    var isConnected: Bool = false
    
    init(peripheral: CBPeripheral, rssi: Int = 0) {
        self.peripheral = peripheral
        self.name = peripheral.name ?? "Unknown Device"
        self.rssi = rssi
    }
}

// MARK: - Heart Rate Monitor Delegate

protocol HeartRateMonitorDelegate: AnyObject {
    func didDiscoverDevice(_ device: HeartRateDevice)
    func didConnectToDevice(_ device: HeartRateDevice)
    func didDisconnectFromDevice(_ device: HeartRateDevice)
    func didReceiveHeartRate(_ reading: HeartRateReading)
    func didFailWithError(_ error: Error)
}

// MARK: - Heart Rate Monitor

class HeartRateMonitor: NSObject {
    
    // Singleton
    static let shared = HeartRateMonitor()
    
    // Core Bluetooth
    private let centralManager = CBCentralManager()
    
    // Devices
    private var discoveredDevices: [String: HeartRateDevice] = [:]
    private var connectedDevice: HeartRateDevice?
    
    // State
    private(set) var isScanning = false
    private(set) var isMonitoring = false
    
    // Buffer for batching
    private var readingBuffer: [HeartRateReading] = []
    private let batchSize = 10
    private var lastEmitTime: Date?
    
    // Delegate
    weak var delegate: HeartRateMonitorDelegate?
    
    // Flutter channel
    private var channel: FlutterMethodChannel?
    
    // MARK: - Initialization
    
    private override init() {
        super.init()
        centralManager.delegate = self
    }
    
    // MARK: - Flutter Channel Setup
    
    func setupFlutterChannel(binaryMessenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "com.titan.app/heart_rate_monitor",
            binaryMessenger: binaryMessenger
        )
        
        channel?.setMethodCallHandler { [weak self] (call, result) in
            switch call.method {
            case "startScan":
                self?.startScan()
                result(nil)
                
            case "stopScan":
                self?.stopScan()
                result(nil)
                
            case "getDiscoveredDevices":
                let devices = self?.getDiscoveredDevices()
                result(devices)
                
            case "connect":
                if let deviceId = call.arguments as? String {
                    self?.connect(deviceId: deviceId)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "Device ID required", details: nil))
                }
                
            case "disconnect":
                self?.disconnect()
                result(nil)
                
            case "getConnectedDevice":
                let device = self?.getConnectedDeviceInfo()
                result(device)
                
            case "startMonitoring":
                self?.startMonitoring()
                result(nil)
                
            case "stopMonitoring":
                self?.stopMonitoring()
                result(nil)
                
            case "getLatestHeartRate":
                let reading = self?.readingBuffer.last?.toDictionary()
                result(reading)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    // MARK: - Scanning
    
    func startScan() {
        guard centralManager.state == .poweredOn else {
            sendError("Bluetooth is not powered on")
            return
        }
        
        discoveredDevices.removeAll()
        isScanning = true
        
        // Scan for devices with heart rate service
        centralManager.scanForPeripherals(
            withServices: [HeartRateServiceUUID.heartRate],
            options: [
                CBCentralManagerScanOptionAllowDuplicatesKey: false
            ]
        )
        
        print("[HeartRateMonitor] Started scanning for HR devices")
    }
    
    func stopScan() {
        centralManager.stopScan()
        isScanning = false
        print("[HeartRateMonitor] Stopped scanning")
    }
    
    func getDiscoveredDevices() -> [[String: Any]] {
        return discoveredDevices.values.map { device in
            return [
                "id": device.peripheral.identifier.uuidString,
                "name": device.name,
                "rssi": device.rssi,
                "isConnected": device.isConnected
            ]
        }
    }
    
    // MARK: - Connection
    
    func connect(deviceId: String) {
        guard let device = discoveredDevices[deviceId] else {
            sendError("Device not found: \(deviceId)")
            return
        }
        
        // Stop scanning before connecting
        stopScan()
        
        centralManager.connect(device.peripheral, options: [
            CBConnectPeripheralOptionNotifyOnDisconnectionKey: true
        ])
        
        print("[HeartRateMonitor] Connecting to: \(device.name)")
    }
    
    func disconnect() {
        guard let device = connectedDevice else { return }
        
        centralManager.cancelPeripheralConnection(device.peripheral)
        connectedDevice = nil
        
        print("[HeartRateMonitor] Disconnected")
    }
    
    func getConnectedDeviceInfo() -> [String: Any]? {
        guard let device = connectedDevice else { return nil }
        return [
            "id": device.peripheral.identifier.uuidString,
            "name": device.name,
            "batteryLevel": device.batteryLevel ?? NSNull()
        ]
    }
    
    // MARK: - Monitoring
    
    func startMonitoring() {
        isMonitoring = true
        readingBuffer.removeAll()
        print("[HeartRateMonitor] Started monitoring")
    }
    
    func stopMonitoring() {
        isMonitoring = false
        emitBuffer()  // Flush remaining readings
        print("[HeartRateMonitor] Stopped monitoring")
    }
    
    // MARK: - Data Processing
    
    private func parseHeartRateMeasurement(data: Data) -> Int {
        // Heart Rate Measurement characteristic format
        // Flags byte: bit 0 = HR format (0 = UINT8, 1 = UINT16)
        //            bit 1-2 = Sensor contact status
        //            bit 3 = Energy expended present
        //            bit 4 = RR intervals present
        
        guard data.count >= 2 else { return 0 }
        
        let flags = data[0]
        let is16Bit = (flags & 0x01) != 0
        
        let bpm: Int
        if is16Bit {
            bpm = Int(data[1]) | (Int(data[2]) << 8)
        } else {
            bpm = Int(data[1])
        }
        
        return bpm
    }
    
    private func parseBatteryLevel(data: Data) -> Int {
        guard data.count >= 1 else { return 0 }
        return Int(data[0])
    }
    
    private func processHeartRate(bpm: Int, source: String) {
        let reading = HeartRateReading(
            bpm: bpm,
            hrv: nil,  // Would need RR intervals for HRV calculation
            source: source,
            timestamp: Date(),
            batteryLevel: connectedDevice?.batteryLevel
        )
        
        readingBuffer.append(reading)
        delegate?.didReceiveHeartRate(reading)
        
        // Check if we should emit batch
        if readingBuffer.count >= batchSize {
            emitBuffer()
        }
    }
    
    private func emitBuffer() {
        guard !readingBuffer.isEmpty else { return }
        
        let batch = readingBuffer.map { $0.toDictionary() }
        
        channel?.invokeMethod("onHeartRateBatch", arguments: [
            "readings": batch
        ])
        
        lastEmitTime = Date()
        readingBuffer.removeAll()
        
        print("[HeartRateMonitor] Emitted batch of \(batch.count) readings")
    }
    
    // MARK: - Error Handling
    
    private func sendError(_ message: String) {
        let error = NSError(domain: "HeartRateMonitor", code: -1, userInfo: [NSLocalizedDescriptionKey: message])
        delegate?.didFailWithError(error)
        channel?.invokeMethod("onError", arguments: message)
    }
}

// MARK: - CBCentralManagerDelegate

extension HeartRateMonitor: CBCentralManagerDelegate {
    
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            print("[HeartRateMonitor] Bluetooth powered on")
            channel?.invokeMethod("onBluetoothStateChanged", arguments: "poweredOn")
            
        case .poweredOff:
            print("[HeartRateMonitor] Bluetooth powered off")
            channel?.invokeMethod("onBluetoothStateChanged", arguments: "poweredOff")
            
        case .unauthorized:
            sendError("Bluetooth permission denied")
            
        case .unsupported:
            sendError("Bluetooth not supported on this device")
            
        default:
            break
        }
    }
    
    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        let deviceId = peripheral.identifier.uuidString
        
        // Update or add device
        var device = HeartRateDevice(peripheral: peripheral, rssi: RSSI.intValue)
        
        if let existing = discoveredDevices[deviceId] {
            device.isConnected = existing.isConnected
        }
        
        discoveredDevices[deviceId] = device
        
        // Notify Flutter
        channel?.invokeMethod("onDeviceDiscovered", arguments: [
            "id": deviceId,
            "name": device.name,
            "rssi": device.rssi
        ])
        
        delegate?.didDiscoverDevice(device)
        
        print("[HeartRateMonitor] Discovered: \(device.name) (\(deviceId))")
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        let deviceId = peripheral.identifier.uuidString
        
        // Update device state
        discoveredDevices[deviceId]?.isConnected = true
        
        // Set as connected device
        if var device = discoveredDevices[deviceId] {
            device.isConnected = true
            connectedDevice = device
        }
        
        // Discover services
        peripheral.discoverServices([
            HeartRateServiceUUID.heartRate,
            HeartRateServiceUUID.batteryService,
            HeartRateServiceUUID.deviceInfoService
        ])
        
        // Notify
        channel?.invokeMethod("onDeviceConnected", arguments: [
            "id": deviceId,
            "name": connectedDevice?.name ?? "Unknown"
        ])
        
        delegate?.didConnectToDevice(connectedDevice!)
        
        print("[HeartRateMonitor] Connected to: \(connectedDevice?.name ?? "Unknown")")
    }
    
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        let deviceId = peripheral.identifier.uuidString
        
        // Update device state
        discoveredDevices[deviceId]?.isConnected = false
        
        // Clear connected device
        if connectedDevice?.peripheral.identifier.uuidString == deviceId {
            let disconnectedDevice = connectedDevice
            connectedDevice = nil
            delegate?.didDisconnectFromDevice(disconnectedDevice!)
        }
        
        // Notify
        channel?.invokeMethod("onDeviceDisconnected", arguments: [
            "id": deviceId
        ])
        
        print("[HeartRateMonitor] Disconnected from: \(deviceId)")
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        sendError("Failed to connect: \(error?.localizedDescription ?? "Unknown error")")
    }
}

// MARK: - CBPeripheralDelegate

extension HeartRateMonitor: CBPeripheralDelegate {
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil else {
            sendError("Service discovery failed: \(error!.localizedDescription)")
            return
        }
        
        guard let services = peripheral.services else { return }
        
        for service in services {
            if service.uuid == HeartRateServiceUUID.heartRate {
                peripheral.discoverCharacteristics(
                    [HeartRateServiceUUID.heartRateMeasurement],
                    for: service
                )
            } else if service.uuid == HeartRateServiceUUID.batteryService {
                peripheral.discoverCharacteristics(
                    [HeartRateServiceUUID.batteryLevel],
                    for: service
                )
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard error == nil, let characteristics = service.characteristics else { return }
        
        for characteristic in characteristics {
            if characteristic.uuid == HeartRateServiceUUID.heartRateMeasurement {
                // Enable notifications for heart rate
                peripheral.setNotifyValue(true, for: characteristic)
                print("[HeartRateMonitor] Subscribed to heart rate notifications")
                
            } else if characteristic.uuid == HeartRateServiceUUID.batteryLevel {
                // Read battery level once
                peripheral.readValue(for: characteristic)
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard error == nil, let data = characteristic.value else { return }
        
        if characteristic.uuid == HeartRateServiceUUID.heartRateMeasurement {
            let bpm = parseHeartRateMeasurement(data: data)
            
            if bpm > 0 && isMonitoring {
                processHeartRate(bpm: bpm, source: connectedDevice?.name ?? "BLE")
                
                // Send real-time update
                channel?.invokeMethod("onHeartRateUpdate", arguments: [
                    "bpm": bpm,
                    "timestamp": Date().timeIntervalSince1970 * 1000
                ])
            }
            
        } else if characteristic.uuid == HeartRateServiceUUID.batteryLevel {
            let battery = parseBatteryLevel(data: data)
            connectedDevice?.batteryLevel = battery
            
            channel?.invokeMethod("onBatteryLevelUpdate", arguments: battery)
            print("[HeartRateMonitor] Battery level: \(battery)%")
        }
    }
}

// MARK: - Flutter Plugin Registration

class HeartRateMonitorPlugin: NSObject, FlutterPlugin {
    
    static func register(with registrar: FlutterPluginRegistrar) {
        HeartRateMonitor.shared.setupFlutterChannel(binaryMessenger: registrar.messenger())
    }
}